Use static_cast instead of C-style casts for non-fundamental types

practicalswift opened the pull request:
A C-style cast is equivalent to try casting in the following order:

1. `const_cast(...)`
2. `static_cast(...)`
3. `const_cast(static_cast(...))`
4. `reinterpret_cast(...)`
5. `const_cast(reinterpret_cast(...))`

By using `static_cast<T>(...)` explicitly we avoid the possibility of an unintentional and dangerous `reinterpret_cast`. Furthermore `static_cast<T>(...)` allows for easier grepping of casts.

For a more thorough discussion, see ["ES.49: If you must use a cast, use a named cast"](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast) in the C++ Core Guidelines (Stroustrup & Sutter).

benma commented on src/net_processing.cpp:
```diff
@@ -2629,7 +2629,7 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman& connman)
 
     BOOST_FOREACH(const CBlockReject& reject, state.rejects) {
-        connman.PushMessage(pnode, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::REJECT, (std::string)NetMsgType::BLOCK, reject.chRejectCode, reject.strRejectReason, reject.hashBlock));
+        connman.PushMessage(pnode, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::REJECT, static_cast<std::string>(NetMsgType::BLOCK), reject.chRejectCode, reject.strRejectReason, reject.hashBlock));
```
maybe `std::string(NetMsgType::BLOCK)` is nicer here.

benma commented on src/primitives/block.h:
```diff
@@ -86,14 +86,14 @@ class CBlock : public CBlockHeader
     {
         SetNull();
-        *((CBlockHeader*)this) = header;
+        *(static_cast<CBlockHeader*>(this)) = header;
```
Why doesn't this need a cont_cast? `this` is `const`, isn't it?

practicalswift replied:
@benma `this` isn't `const` in the constructor:

> The type of this in a member function of class X is X* (pointer to X). If the member function is cv-qualified, the type of this is cv X* (pointer to identically cv-qualified X). Since constructors and destructors cannot be cv-qualified, the type of this in them is always X*, even when constructing or destroying a const object. ([cppreference.com](http://en.cppreference.com/w/cpp/language/this))

benma replied:
Of course, `this` is only const if the function is const. My bad.

benma commented:
utACK 1a3c1e24cbce60910843f434dc6c489d03eaed1d

practicalswift committed:
Use static_cast instead of C-style casts for non-fundamental types

A C-style cast is equivalent to try casting in the following order:

1. const_cast(...)
2. static_cast(...)
3. const_cast(static_cast(...))
4. reinterpret_cast(...)
5. const_cast(reinterpret_cast(...))

By using static_cast<T>(...) explicitly we avoid the possibility
of an unintentional and dangerous reinterpret_cast. Furthermore
static_cast<T>(...) allows for easier grepping of casts.

benma commented:
utACK 6090e51b5d3f8024197a31e6a5a50b812ba10aca

MarcoFalke commented:
utACK 6090e51b5d3f8024197a31e6a5a50b812ba10aca does not change the binariy on my platform

practicalswift commented:
Rebased!

practicalswift commented:
Rebased!

practicalswift commented:
Rebased! :-)

practicalswift commented:
Rebased! :-)

practicalswift commented:
Rebased!

practicalswift commented:
Anyone willing to review? ACK or NACK? :-)

kekimusmaximus commented:
Why not go the extra mile and also add -Wold-style-cast, otherwise you'll be back fixing it a week later.

ryanofsky reviewed (commented):
utACK 9ad6746ccd6dc31141fd0144686b641e31bf626b

MarcoFalke commented:
re-reviewed

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-utACK 9ad6746ccd6dc31141fd0144686b641e31bf626b
-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJacgWeAAoJENLqSFDnUoslDH0P/0t30sv4UootDVMUutXKfnsK
mHKRDOKMbWLYQonJ93x89Oll+U+r+B7m7CzlxqxUNI2Zd6DETOprwLAG8F1yKHez
JUjjndpfP/6wgzyv+YEJPehFL2vVSu3Di0j82qk/1vSaYHA94bh6Y+pRnVE65951
YdpuD0scPfXIMuR93swiGpw8ttFEqZVcGtjzWlwgKO23cdJaIlZatmpi1Mw/eQea
pqhol7AgtzZeEnISYDRVcObPxVzUabGDkVXUjafJK14enMLpg6ddjV4tldLCtVfR
J5yEfxGGTcYLm5TYpvYGQS58GWZSdbUm22HPDCcjYZW+NVvXRcvsaIYPE+BZJ/u/
q9jEs3wAP7lhTp0XiFmJ6UFlyrkVBopGmNI1KPWtsxdvjoouAH8XpO3RsB6/pNZq
nLZcvg/QNSbJ5ypMFshcZ2iJjVtgmIuwC/kceAuy3nK0EkK9BFUd9qS3dzdAbYiZ
aYfZjn6fu0PWxItMzEMActleYwAdhQLjc6tX35tSlVE0ZOSJx0FDWi6dIPyzlX6L
jWGKpMCm2m/Xt19av4IS9AHk+xoIGsQrgaWt+7TYc7vssR/dlw8IXsKI2D68AEJE
IXxh6HGNeBu/2cbTQ4bwA4CtAaE5sS44EdbxbduF1mAPhUJEoywP51F9VkNAqaQQ
64GCcFS6tzqUfliIrLGl
=Gwy1
-----END PGP SIGNATURE-----

practicalswift commented:
@kekimusmaximus That would be nice but I think there might be some instances of C-style casts for non-fundamental types in our dependencies, no? :-)

ryanofsky commented:
Seems like this could be merged: obvious change, has 3 acks.

Sjors commented on src/qt/bitcoin.cpp:
```diff
@@ -709,7 +709,7 @@ int main(int argc, char *argv[])
             app.requestInitialize();
 #if defined(Q_OS_WIN) && QT_VERSION >= 0x050000
-            WinShutdownMonitor::registerShutdownBlockReason(QObject::tr("%1 didn't yet exit safely...").arg(QObject::tr(PACKAGE_NAME)), (HWND)app.getMainWinId());
+            WinShutdownMonitor::registerShutdownBlockReason(QObject::tr("%1 didn't yet exit safely...").arg(QObject::tr(PACKAGE_NAME)), static_cast<HWND>(app.getMainWinId()));
```
Maybe unrelated, but I'm trying to make a Gitian Windows build for a branch based off of master and it throws:
```
qt/bitcoin.cpp: In function ‘int main(int, char**)’:
qt/bitcoin.cpp:709:173: error: invalid static_cast from type ‘WId {aka unsigned int}’ to type ‘HWND’
             WinShutdownMonitor::registerShutdownBlockReason(QObject::tr("%1 didn't yet exit safely...").arg(QObject::tr(PACKAGE_NAME)), static_cast<HWND>(app.getMainWinId()));                                                                                                                                                                       ^
```

Sjors replied:
Just noticed #12386...
